Blocks E2E: Eliminate and prevent side effects through DB snapshots

WunderBart opened the pull request:
**Related Issue:** https://github.com/woocommerce/woocommerce/issues/45347  
**Closes:** https://github.com/woocommerce/woocommerce/issues/46329

## Overview

This PR addresses side effects in our Blocks E2E tests by ensuring each test operates in a more isolated WordPress environment. For a detailed explanation of WordPress E2E test side effects and their implications, please refer to the issue described in https://github.com/woocommerce/woocommerce/issues/46329. We have enhanced test isolation by introducing DB snapshots. This is achieved by using the [wp db](https://developer.wordpress.org/cli/commands/db/) API to export the database post-initial setup and re-importing this baseline state before each test, ensuring each test starts from a clean slate and maintains both isolation and independence.

### Limitations and benefits of the approach

Our primary modification is the exclusive use of the `beforeEach` hook, which necessitates avoiding `beforeAll`, `afterAll`, and `afterEach` hooks to prevent interference with the database state resets between tests. This approach, though seemingly restrictive, streamlines the testing process by removing the need for cleanup during the teardown phase. Test setup now focuses solely on preparation, supported by an `eslint` rule that mandates using the `beforeEach` hook. Importing a database snapshot for each test introduces a minimal delay—about 1 second to the setup time—but this is balanced by eliminating nearly all teardown-related code.

_I tried to utilize raw Docker commands, which reduced this delay to under 200 milliseconds in early testing. However, that approach requires much more maintenance, so I'll pursue further reductions and optimizations in a subsequent PR if necessary._ 

### Future-Proofing against side effects

With these changes implemented, future tests are inherently shielded from side effects, enhancing our testing environment's stability and reliability.

## Next steps

Having controlled the side effects, we are now looking to consolidate all configurations into a single suite to optimize resource utilization through proportional sharding. To maintain clarity, we will not remove the `*.side_effects.*` filename tag in this PR. The removal of this tag will be handled in the upcoming follow-up PR. You can review the draft PR for this next phase here: https://github.com/woocommerce/woocommerce/pull/46409.

## Testing instructions

This is a backward-compatible refactor; therefore, functionality should remain as it was. Besides ensuring the CI tests pass, we encourage running some blocks tests locally to verify that all components function properly.

WunderBart committed:
try db reset in page teardown

WunderBart committed:
move reset to setup step

WunderBart committed:
Use wp db cli

WunderBart committed:
Fix global setup
That part is overriding the logged in user state and wipes out the nonce and rootUrl fields.

WunderBart committed:
Try importing instead from a generated dump

WunderBart committed:
Revert "Try importing instead from a generated dump"

This reverts commit 987dc471c9dab49dd14aa8a96be43418454133a1.

WunderBart committed:
Revert "Revert "Try importing instead from a generated dump""

This reverts commit c8d008cb2040c03c4cebc725734856fc5a472017.

WunderBart committed:
Don't bypass visitSiteEditor so that the Welcome Guide is closed

gigitux committed:
use createNewPost

WunderBart committed:
Revert "Revert "Revert "Try importing instead from a generated dump"""

This reverts commit 26842735822e95252fbe1cc63b026db4c1949f16.

WunderBart committed:
Merge remote-tracking branch 'origin' into try/e2e-db-reset

WunderBart committed:
Merge remote-tracking branch 'origin' into try/e2e-db-reset

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into try/e2e-db-reset

gigitux committed:
[Blocks]: Fix E2E tests (#46344)

* Load local pickup enabled setting as bool not string

* Add changelog

* Remove unused import

* Update plugins/woocommerce-blocks/assets/js/extensions/shipping-methods/pickup-location/utils.ts

Co-authored-by: Niels Lange <info@nielslange.de>

* Use strict equality test

* try now

* try now

* add a new sql file

* remove default.sql

* fix welcome guide tour

* fix E2E tests

* Fix the template revert tests
...where the template is unreachable due to pagination.

* Add changelog entry

* improve logic to closeWelcomeGuideModal

* fix cart checkout tests

* improve flakiness

* improve flakiness

* fix flaky test

* fix company field

* fix E2E test

* fix E2E tests

---------

Co-authored-by: Thomas Roberts <thomas.roberts@automattic.com>
Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
Co-authored-by: Niels Lange <info@nielslange.de>
Co-authored-by: Bart Kalisz <bartlomiej.kalisz@gmail.com>

WunderBart committed:
Replace beforeAll w/ beforeEach + remove all after* hooks

WunderBart committed:
Fix broken hooks

WunderBart committed:
Activate plugins via requestUtils API

gigitux committed:
[Blocks - E2E]: Add `playwright/no-hooks` ESlint rule (#46432)

add ESLint configuration

WunderBart committed:
Clean up console logs

WunderBart committed:
Remove obsolete language setup steps

WunderBart committed:
Remove more unnecessary setup steps

WunderBart committed:
Remove even more obsolete setup steps

WunderBart committed:
tmp: add the LYS fix

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into try/e2e-db-reset

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/checkout-block.merchant.block_theme.spec.ts:
```diff
@@ -78,7 +78,7 @@ test.describe( 'Merchant → Checkout', () => {
+		test.beforeEach( async ( { browser } ) => {
 			const page = await browser.newPage();
 			await page.goto(
 				`${ process.env.WORDPRESS_BASE_URL }/?setup_terms_and_privacy`
```
This call relies on the helper plugin `woo-test-helper.php` which is still located at `plugins/woocommerce-blocks/tests/mocks/woo-test-helper/woo-test-helper.php`. What are your thoughts on relocating it to `plugins/woocommerce-blocks/tests/e2e/plugins`? If we relocate it, we should also remove it from `plugins/woocommerce-blocks/.wp-env.json`.

WunderBart replied:
Nice catch, I missed this one! I'll move it to the folder where the other plugins are.

WunderBart replied:
OTOH, this seems to be a bit different case from the plugins in the e2e/plugins folder. They're installed but inactive until requested on a per-test basis. I'm not saying the woo-test-helper should not be there; it seems like it needs to be active by default, which is a bit different use case. Let's leave it for a follow-up PR, since this one's already pretty big! 🙈

nielslange replied:
I just created https://github.com/woocommerce/woocommerce/issues/46745, to take care of relocating the helper plugin `woo-test-helper.php` once this PR had been merged.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into try/e2e-db-reset

WunderBart committed:
Try stabilizing the company field test

WunderBart committed:
Use canvas param instead of manually entering edit mode

WunderBart committed:
Remove double site editor redirect

nielslange reviewed (commented):
Thanks for working on this, @WunderBart. I left a comment regarding a helper plugin, which I introduced a while ago, but which wasn't located within the tests.

WunderBart committed:
Revert "Use canvas param instead of manually entering edit mode"

This reverts commit 5e6cc1715490ed333b2e363e1cdcefe4e8696683.

WunderBart committed:
Revert "Remove double site editor redirect"

This reverts commit 69a57a82a86ab0fe39002bc59838e6a63d118298.

WunderBart committed:
Fix flaky products sorting test

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into try/e2e-db-reset

gigitux committed:
Blocks: Fix ESLint errors (#46595)

fix eslint error

opr commented on plugins/woocommerce-blocks/tests/e2e/tests/page-content-wrapper/page-content-wrapper.block_theme.side_effects.spec.ts:
```diff
@@ -36,17 +36,17 @@ const userText = 'Hello World in the page';
 			await admin.visitAdminPage( 'edit.php?post_type=page' );
+
 			await page.getByLabel( `“${ template.title }” (Edit)` ).click();
+			await page.waitForLoadState();
```
Are you waiting for the load state just to be sure navigation has completed? A comment here would be useful, I spent a couple of mins working out what this line did

WunderBart replied:
I don't think this line is actually necessary since `closeWelcomeGuideModal` waits for the `data` API to be available. Can we remove it? (🏓 @gigitux since you added this line)

gigitux replied:
Let's remove it!

WunderBart replied:
Removed in https://github.com/woocommerce/woocommerce/pull/46125/commits/af254f4a2d80e581e5b0df0dafc8e255daf66745

opr reviewed (commented):
This is looking really good @WunderBart thanks a lot for working on this. Restoring from a db before each test is a good idea, despite feeling like it's a lot of work to do before each test, it is probably way more work than re-running tests in CI.

I am impressed by how much code you've removed.

I wonder if it would be good to add some documentation with guidelines about writing E2E tests to the repo here - https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks/docs/internal-developers/testing

It can be done as a follow-up but I think making sure folks are aware of the patterns you've introduced here is important.

gigitux committed:
Fix ESLint errors (#46626)

* Fix ESLint errors

* Add changefile(s) from automation for the following project(s): woocommerce-blocks

---------

Co-authored-by: github-actions <github-actions@github.com>

gigitux committed:
Merge branches 'try/e2e-db-reset' and 'trunk' of github.com:woocommerce/woocommerce into try/e2e-db-reset

WunderBart committed:
Merge remote-tracking branch 'origin' into try/e2e-db-reset

WunderBart committed:
Remove obsolete waitForLoadState call

WunderBart committed:
Merge remote-tracking branch 'origin' into try/e2e-db-reset

WunderBart committed:
Merge remote-tracking branch 'origin' into try/e2e-db-reset

nielslange reviewed (commented):
The changes look great, @WunderBart. However, I see that `plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-checkout-fields-plugin.php` has a merge conflict and two tests in [Side Effects [2/5]](https://github.com/woocommerce/woocommerce/actions/runs/8799923100/job/24151902418?pr=46125#logs) are currently failing. I briefly looked up the failing test and somehow, the welcome guide model does not get closed.

WunderBart committed:
Merge remote-tracking branch 'origin' into try/e2e-db-reset

WunderBart committed:
Merge remote-tracking branch 'origin' into try/e2e-db-reset

WunderBart committed:
Fix the Page Content Wrapper test
- Refactor out the flaky closeWelcomeGuideModal
- Remove the obsolete teardown steps

WunderBart committed:
Fix classic template tests

WunderBart committed:
Make linter happy

Aljullu commented on plugins/woocommerce-blocks/tests/e2e/tests/single-product-template/single-product-template.block_theme.spec.ts:
```diff
@@ -41,6 +41,7 @@ const products = [
 
 for ( const { classes, product, frontendPage } of products ) {
+	// eslint-disable-next-line playwright/valid-describe-callback
 	test.describe( `The Single Product page of the ${ product }`, () =>
```
Instead of disabling `playwright/valid-describe-callback`, would it be possible to wrap the describe callback in `{}` so it doesn't return anything (and then, the describe callback is valid)?

WunderBart replied:
Nice catch! That describe should definitely not be returning that `test`! 😄

WunderBart replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/46125/commits/d0c5c334b3006218c9117b40540b1bb3690d05c2

Aljullu commented on plugins/woocommerce/changelog/46626-fix-eslint-errors:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: dev
+Comment: Fix ESLint errors
```
Not a big deal, but I think this PR does much more than this. :smile: It might be good to update the comment to align with the PR contents.

WunderBart replied:
Haha definitely! That must have been merged along with #46626 that fixed the linting errors. Will update!

WunderBart replied:
Updated in https://github.com/woocommerce/woocommerce/pull/46125/commits/243d44d09e99e988319388fcd1859fd1857aa414

Aljullu reviewed (approved):
Awesome job with this PR, @WunderBart! Thanks for all the work improving e2e tests. :clap:  I left a couple of minor comments, but LGTM.

WunderBart committed:
Fix the invalid describe block

WunderBart committed:
Update changelog entry
